README.md
Tropo AGItate
==============

Tropo AGItate lets you run Asterisk AGI applications on the Tropo cloud communications platform.

Description
-----------

Provides a script to emulate the Fast Asterisk Gateway Interface [FastAGI]([http://www.voip-info.org/wiki/view/Asterisk+FastAGI "FastAGI") protocol on [Tropo](http://tropo.com "Tropo"). While most of the development and testing has been done using [Adhearsion](http://adhearsion.com "Adhearsion"), we stayed with the proper Asterisk AGI implementation. Therefore this script should work with any FastAGI server including [PHPAGI](http://phpagi.sourceforge.net/ "PHPAGI") and [Asterisk-Java](http://asterisk-java.org/ "Asterisk Java"). We will be doing more testing with alternative frameworks soon and reporting back.

Installation
-----------

This project allows you to control the Tropo AGI via an Adhearsion dialplan context.

* Create a free account at [Tropo](http://tropo.com "Tropo").

* Install [Adhearsion 0.8.6+](http://adhearsion.com "Adhearsion"). You can install this on your local computer.

<pre>
  gem install adhearsion
</pre>

* Create a simple dialplan, or use some of the examples provided

<pre>
  tropo_agi {
    play "Wow. My first app! Tropo really is this easy!"
  }
</pre>

* Ensure your Adhearsion app is running and has a public IP. You may need to port forward 4573 on your firewall.

* Create your Tropo application

Instructions may be found [here](https://www.tropo.com/docs/scripting/quickstart.htm "Quick Start").

* Set your Tropo app to use the [lib/tropo-agitate.rb](https://github.com/tropo/tropo-agitate/raw/master/lib/tropo-agitate.rb) script

  * Load the file directly to the Github file location [lib/tropo-agitate.rb](https://github.com/tropo/tropo-agitate/raw/master/lib/tropo-agitate.rb)
  * Host the 'tropo-agitate.rb' yourself and provide a public URL for access
  * Add it to your Tropo FTP/WebDAV account

* Modify the configuration file and post it to your Tropo FTP/WebDAV account

  * First set your configuration settings in the [tropo_agi_config/tropo_agi_config.yml](https://github.com/tropo/tropo-agitate/raw/master/tropo_agi_config/tropo_agi_config.yml "tropo_agi_config.yml") file
  * Then upload to your [Tropo FTP/WebDAV](https://www.tropo.com/docs/scripting/tropohosting.htm) account placing in root/www/tropo_agi_config

* Use a SIP client (like Blink for Mac) and dial the SIP Voice account listed under your application, may use Skype too!

* Happy Tropo-ing!

Placing Outbound Calls
----------------------

Tropo supports placing outbound calls with the [Session API](https://www.tropo.com/docs/scripting/sessions.htm). AGItate allows for receiving a predefined set of parameters to place the call and tag it so you may associate it to your request once it reaches your AGI server. The parameters supported are:

* destination

  The destination to make a call to or send a message to. This may currently take one of the following forms:

  14155551212 - The telephone number to dial with the country code.
  sip:username@doamin.com - The SIP URI to dial
  username - The IM or Twitter user name.

  Some IM networks like Google Talk and Live Messenger include a domain as part of the user name. For those networks, include the domain: username@gmail.com

  When making a voice call, you can specify dialing options as part of the number:

  You can also list multiple phone numbers or SIP addresses (or both!) as a comma separated list

* caller_id

  The Caller ID for the session's origin. For example, if the number (407)555-1212 called (407)555-1000, the *1212 number would be the callerID. This also applies to IM account names; if IM account 'tropocloud' sends a message to 'foobar123', the callerID would be 'tropocloud'.

  The callerID can be manually set to a specific number; for voice calls, this can be any valid phone number, though for SMS and IM it must be a number/ID assigned to your account.

* channel

  Channel tells Tropo whether the call is "voice" or "text".

* network

  Network is used mainly by the text channels; values can be SMS when sending a text message, or a valid IM network name such as AIM, GTALK, MSN, JABBER, TWITTER and YAHOO. For IM network, you must have an IM account linked in your app. For example, if you try to send to AIM when you don't have an AIM username included in your app, your app will fail.

* timeout

  Timeout only applies to the voice channel and determines the amount of time Tropo will wait - in seconds - for the call to be answered before giving up.

* tropo_tag

  An arbitrary unique identifier that you may use to identify the call once it is placed and passed to your AGI server. This value will appear in the tropo\_headers variable received at the beginning of the request to your AGI server in the JSON as tropo\_tag.

You may then invoke a call request via HTTP as follows:

* GET
<pre>
  http://api.tropo.com/1.0/sessions?action=create&token=TOKEN&destination=NUMBER&caller\_id=CALLINGNUMBER&tropo_tag=1234
</pre>

* POST
<pre>
  http://api.tropo.com/1.0/sessions
  <code>
  <session>
    <token>YOUR_TOKEN</token>
    <var name="destination" value="4155551212" />
    <var name="caller_id" value="7146432997" />
    <var name="tropo_tag" value=1234 />
  </session>
  </code>
</pre>

Supported Adhearsion & AGI Methods
----------------------------------

Refer to the wiki [Supported Adhearsion and Asterisk Gateway Interface (AGI) Commands](http://github.com/tropo/tropo-agitate/wiki/Supported-Adhearsion-&-AGI-Methods) page.

Asterisk Sound Files Available
------------------------------

Refer to the wiki [Asterisk Core Sounds Available](http://github.com/tropo/tropo-agitate/wiki/Built-In-Asterisk-Sound-Files) page. This will also let you know how to add your own custom sound files.

Adhearsion Dialplan Examples
----------------------------

These Adhearsion dialplan examples are also available in the 'examples' directory of this project.

<pre>
  # The main AGI entry point
  tropo_agi {
    answer
    +hello_world
    hangup
  }


  # Invokes the native Tropo ask for Speech Recognition / ASR
  # Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
  ask_example {
    result = execute 'ask', { :prompt  => 'Please enter your zip code',
                              :choices => 'zipcode([5 DIGITS])' }.to_json
    # Get rid of the '200 result =' and then parse the JSON
    response = JSON.parse(result[11..-1])
    ahn_log.debug response
  }

  # We have lots of Asterisk sounds!
  asterisk_sounds_example {
    play 'tt-monkeys'
  }

  # You may dial other SIP addresses, or telephone numbers as you like
  # Formats may be:
  #  SIP/
  #  sip:
  #  tel:+
  # Related Tropo method: https://www.tropo.com/docs/scripting/transfer.htm
  dial_example {
    dial 'sip:9991443146@sip.tropo.com'
  }

  hello_world {
    play 'tt-monkeys'
  }

  # Works as input should, only accepting DTMF
  # Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
  input_example {
    postal_code = input 5, :play         => 'Please enter your post code?',
                           :timeout      => 2

    ahn_log.postal_code.debug postal_code
    play "You entered"
    say_digits postal_code
  }

  # Find out if this is a Tropo session or an Asterisk one
  is_tropo? {
    if type == 'TROPO'
      play "Yippeee! It is a Tropo call!"
    else
      play "No, this is a good old Asterisk call"
    end
  }

  # Related Tropo method: https://www.tropo.com/docs/scripting/say.htm
  say_digits_example {
    say_digits '12345'
  }

  # Uses the native Tropo say method for Speech-Synthesis/TTS, will not play Asterisk sound files like play will
  # Related Tropo method: https://www.tropo.com/docs/scripting/say.htm
  say_example {
    execute 'say', { :prompt => 'I like to have weasels in my cloud.', :voice => 'simon' }.to_json
  }

  # Supports STREAM FILE with and without escape digts
  # Note, any digits at the end of the string will always be treated as escape digits
  stream_file_example {
	raw_response "STREAM FILE tt-monkeys 1234567890*#"
    raw_response "STREAM FILE tt-monkeys"
  }
 
  # If this is a Tropo call, then all of the SIP headers for the session are available
  show_call_data {
    ahn_log.tropo_headers_str.debug tropo_headers
    tropo_headers = JSON.parse self.tropo_headers
    ahn_log.tropo_headers_hash.debug tropo_headers
    play "The content type is " + tropo_headers['Content-Type']
    +asterisk_sounds_example
  }

  # Monitor and Mixmonitor behave the same, may also be invoked as startCallRecording
  # Related Tropo method: https://www.tropo.com/docs/scripting/startcallrecording.htm
  monitor_example {
    play 'About to start call recording'
    execute 'monitor', { :uri                 => 'http://tropo-audiofiles-to-s3.heroku.com/post_audio_to_s3?filename=voicemail.mp3',
                         :format              => 'mp3',
                         :method              => 'POST',
                         :transcriptionOutURI => 'mailto:jsgoecke@voxeo.com' }.to_json
    play 'Call recording started!'
    play 'Thats it folks!'
    execute 'monitor_stop', ''
    play 'Recording stopped!'
  }

  # Related Tropo method: https://www.tropo.com/docs/scripting/record.htm
  record_prompt_example {
    play 'Please record after the beep'
    record 'http://tropo-audiofiles-to-s3.heroku.com/post_audio_to_s3?filename=voicemail.mp3',
           :silence => 5,
           :maxduration => 120
  }

  # Allows you to set and retrieve variables on the session
  # These also get passed as custom SIP headers, prepended with 'x-tropo' when you dial/transfer a call
  variables_example {
    set_variable('foobar', 'green')
    ahn_log.debug get_variable('foobar')
  }

  # We have lots of Asterisk sounds!
  asterisk_sounds_example {
    play 'tt-monkeys'
  }

  # Allows you to send standard DTMF digits
  send_dtmf_example {
    dtmf '1234567890#*'
  }

  # Shows how to set the voice option for speech synthesis/TTS
  set_voice_example {
    play 'Hello world!'
    execute "voice", "simon"
    play 'Hello world!'
    execute "voice", "default"
    play 'Hello world!'
  }

  # Shows how to set the voice option for speech synthesis/TTS
  set_recognizer_example {
    execute "voice", "carmen"
    execute "recognizer", "es-es"
    result = execute 'ask', { :prompt  => 'Por favor, ingrese su código postal',
                              :choices => 'zipcode([5 DIGITS])' }.to_json
    # Get rid of the '200 result =' and then parse the JSON
    response = JSON.parse(result[11..-1])
    ahn_log.debug response
  }

  # Start menu example
  # Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
  menu_example {
    menu 'Welcome. For Shipments press 1. For ordering press 2. For a representative push 3 or for Spanish press 4', :timeout => 8.seconds, :tries => 3 do |link|
      link.shipment_status  1
      link.ordering         2
      link.representative   3
      link.spanish          4
      link.employee         500..599

      link.on_invalid { play 'invalid' }

      link.on_premature_timeout do |str|
        play 'sorry'
      end

      link.on_failure do
        play 'goodbye'
        hangup
      end
    end
  }

  shipment_status {
    play 'I surely do not know your shipment status.'
  }

  ordering {
    play 'Go somewhere else and order.'
  }

  representative {
    play 'No representatives here.'
  }

  spanish {
    play 'e 2 brutus?'
  }

  employee {
    play "The person at"
    say_digits extension
    play "went home for the day."
  }
  # End menu example
</pre>


RELEASE-NOTES.md
Tropo AGItate - Release Notes
=============================

v0.2.4
* Make AGItate compatible with Ruby 1.9

v0.2.3
------
* Fix bug where dials with an empty duration string would result in a 0-second call (thanks John Dyer for finding/fixing this)

v0.2.2
------
* Ask now handles the options hash propery respecting symbols as keys

v0.2.1
------
* Allow application arguments to be quoted as a single string like: "arg1,arg2,arg3".  This is allowed by Asterisk and is generated by Asterisk.NET.
* Since the above change breaks using Playback for TTS strings, introduce a Tropo-specific "Say" application: EXEC Say {"prompt": "Hello, LSRC!"}

v0.2.0
------
* Add unit tests to cover all AGI primitives documented here: https://wiki.asterisk.org/wiki/display/AST/AGI+Commands
* Overhaul the options parsing to be more exactly in line with Asterisk
* Unify error handling and protocol response messages
* Add lots of unit tests to emulate behavior seen from various AGI clients
* Return sane error messages for applications that do not make sense on Tropo: DATABASE GET/PUT, SET CONTEXT/EXTENSION/PRIORITY, TDD MODE, SEND IMAGE/TEXT, RECEIVE TEXT/CHAR etc.
* Add support for most remaining AGI primitives (exceptions: The SPEECH primitives, SAY ALPHA/DATE/DATETIME etc. are not yet implemented)

v0.1.9
------

* Make unit tests more idempotent	
* Make unit tests less magical	
* Add support for Flexmock	
* Rudimentary answering machine detection	

v0.1.8
------

* Add unit test for outbound callerID functionality	
* Provide has_key? method to ensure proper handling	
* Make sure header keys are strings	
* No need to work with a copy of the channel vars now	
* Make sure to encode as JSON in case we hit a nested channel variable	
* Avoid exception if no input is received	
* Work around Tropo Ruby shim bug with _parseTime
* Better backtrace logging

v0.1.7
------

* Add framework for "magic" channel variables. This supports things like CALLERID(all) vs. CALLERID(name) vs. CALLERID(num) that all have overlap in Asterisk. We now try to do the right thing when setting or reading each variation. More special variables can be easily added.
* Enhanced Dial compatibility: Allow setting the CallerID on outbound calls, set DIALSTATUS based on Tropo response
and clean up parsing of dial string
* Set the default AGI port if unspecified in the YAML
* Update to RSpec 2
* Allow detecting the Tropo dialed number for incoming calls (agi_dnid)
* Fix fatal missing error on SIP failover failure
* Update unit tests for new functionality; fix broken unit tests
* Rspec tests now require JRuby v1.5.x or better


spec/spec_helper.rb
$LOAD_PATH.unshift(File.dirname(__FILE__))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'agi_config'))

%w(rubygems rspec fakeweb eventmachine flexmock tropo tropo-agitate yaml).each { |lib| require lib }
# em-spec/rspec Out for now since it is not Rspec 2.x compat

RSpec.configure do |config|
  config.mock_with :flexmock
  config.filter_run :focus => true
  config.run_all_when_everything_filtered = true
end

# NOTE!
# TESTING REQUIRES JRUBY SINCE WE HAVE TO CREATE A JAVA HASHMAP IN THE TESTS


spec/tropo-agi_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/spec_helper')

describe "TropoAGI" do

  before(:all) do
    # These tests are all local unit tests
    FakeWeb.allow_net_connect = false
    
    # Register where we expect our YAML config file to live
    FakeWeb.register_uri(:get, "http://hosting.tropo.com/49767/www/tropo_agi_config/tropo_agi_config.yml", 
                         :body => File.open('tropo_agi_config/tropo_agi_config.yml').read)
                           
    # Register the hosted JSON file  
    FakeWeb.register_uri(:get, "http://hosting.tropo.com/49767/www/audio/asterisk_sounds/asterisk_sounds.json", 
                         :body => '{"tt-monkeys":"tt-monkeys.gsm"}')
                           
    @current_call = CurrentCall.new
    @tropo_agi = TropoAGI.new(@current_call, CurrentApp.new)
  end
  
  it "should create a TropoAGI object" do
    @tropo_agi.instance_of?(TropoAGI).should == true
  end
  
  it "should create a properly formatted initial message" do
    agi_uri  = URI.parse @tropo_agi.tropo_agi_config['agi']['uri_for_local_tests']
    message  = @tropo_agi.initial_message(agi_uri.host, agi_uri.port, agi_uri.path[1..-1])
    @initial_message = <<-MSG
agi_network: yes
agi_network_script: #{agi_uri.path[1..-1]}
agi_request: agi://#{agi_uri.host}:#{agi_uri.port}#{agi_uri.path}
agi_channel: TROPO/#{@current_call.id}
agi_language: en
agi_type: TROPO
agi_uniqueid: #{@current_call.id}
agi_version: tropo-agi
agi_callerid: #{@current_call.callerID}
agi_calleridname: #{@current_call.callerName}
agi_callingpres: 0
agi_callingani2: 0
agi_callington: 0
agi_callingtns: 0
agi_dnid: 1000
agi_rdnis: unknown
agi_context: #{agi_uri.path[1..-1]}
agi_extension: 1
agi_priority: 1
agi_enhanced: 0.0
agi_accountcode: 
agi_threadid: #{@current_call.id}

MSG
    message.should == @initial_message
  end
  
  it "should parse arguments stripping quotes" do
    result = @tropo_agi.parse_args('"Hello LSRC!"')
    result[0].should == "Hello LSRC!"
    
    result = @tropo_agi.parse_args('"{"prompt":"hi!","timeout":3}"')
    result.should == { "timeout" => 3, "prompt" => "hi!"}
    
    result = @tropo_agi.parse_args('"1234","d",""')
    
    result[0].should == '1234'
    result[1].should == 'd'
    result[3].should == nil
  end
  
  it "should strip quotes from a string" do
    @tropo_agi.strip_quotes('"foobar"').should == 'foobar'
  end
  
  it "should handle commas in non JSON args" do
    command = @tropo_agi.parse_command('EXEC playback "Hello, LRSC!"')
    command.should == { :action => "exec", :command => "playback", :args => ["Hello, LRSC!"] }
  end
  
  it "should extract the appropriate commands from AGI" do
    command = @tropo_agi.parse_command('ANSWER')
    command.should == { :action => "answer" }
    
    command = @tropo_agi.parse_command('HANGUP')
    command.should == { :action => "hangup" }
    
    command = @tropo_agi.parse_command('EXEC playback "Hello LRSC!"')
    command.should == { :action => "exec", :command => "playback", :args => ["Hello LRSC!"] }
        
    command = @tropo_agi.parse_command('EXEC ask "{"prompt":"hi!","timeout":3}"')
    command.should == { :command => "ask", :action => "exec", :args => { "timeout" => 3, "prompt" => "hi!"} }
    
    command = @tropo_agi.parse_command('EXEC Dial "sip:jsgoecke@yahoo.com","",""')
    command.should == { :command => "dial", :action => "exec", :args => ["sip:jsgoecke@yahoo.com", "", ""] }
    
    command = @tropo_agi.parse_command('EXEC MeetMe "1234","d",""')
    command.should == { :command => "meetme", :action => "exec", :args => ["1234", "d", ""] }

    command = @tropo_agi.parse_command('SET CALLERID "9095551234"')
    command.should == { :command => "callerid", :action => "set", :args => ["9095551234"] }
    
    command = @tropo_agi.parse_command('SET MYVAR "foobar"')
    command.should == { :command => "myvar", :action => "set", :args => ["foobar"] }
    
    command = @tropo_agi.parse_command('GET VARIABLE "myvar"')
    command.should == { :command => "variable", :action => "get", :args => ["myvar"] }
    
    command = @tropo_agi.parse_command('EXEC monitor "{"method":"POST","uri":"http://localhost"}"')
    command.should == { :command => "monitor", :action => "exec", :args => { 'method' => 'POST', 'uri' => 'http://localhost' } }
  end
  
  it "should execute the command" do
    command = @tropo_agi.execute_command('EXEC MeetMe "1234","d",""')
    command.should == "200 result=0\n"

    command = @tropo_agi.execute_command('SET CALLERID "9095551234"')
    command.should == "200 result=0\n"
    
    command = @tropo_agi.execute_command('SET CALLERIDNAME "John Denver"')
    command.should == "200 result=0\n"
    
    command = @tropo_agi.execute_command('GET VARIABLE "CALLERIDNAME"')
    command.should == "200 result=1 (John Denver)\n"
    
    command = @tropo_agi.execute_command('SET VARIABLE FOOBAR "green"')
    command.should == "200 result=0\n"
    
    command = @tropo_agi.execute_command('GET VARIABLE "FOOBAR"')
    command.should == "200 result=1 (green)\n"
    
    command = @tropo_agi.execute_command('EXEC monitor "{\"method\":\"POST\",\"uri\":\"http://localhost\"}"')
    command.should == "200 result=0\n"
    
    command = @tropo_agi.execute_command('EXEC mixmonitor "{\"method\":\"POST\",\"uri\":\"http://localhost\"}"')
    command.should == "200 result=0\n"
    
    command = @tropo_agi.execute_command('EXEC sipgetheader "CALLERIDNAME"')
    command.should == "200 result=0\n"
  end
end

lib/tropo.rb
# For testing outside of Tropo we need to mock the global methods
# for logging and call management.
module Tropo
  def log(val)
    #STDERR.puts val
  end

  def show(val)
    log("====> #{val} <====")
  end

  def call(destination, options)
    raise "Unimplemented!"
  end

  class TropoEvent
    attr_accessor :name, :recordURI, :choice, :attempt, :value
    # @value is sometimes set to an instance of TropoCall
  end

  # Here for testing outside of Tropo, we need to mock the return on ask
  class AskResponse
    class Choice
      def concept
        'zipcode'
      end

      def confidence
        '10.0'
      end

      def interpretation
        '94070'
      end

      def tag
        nil
      end
    end

    attr_reader :choice
    attr_accessor :value

    def initialize
      @choice = Choice.new
      @value = '94070'
    end
  end

  # Here for testing outside of Tropo, so we mock the $currentCall object
  class CurrentCall
    attr_reader :value
    attr_reader :isActive
    attr_reader :state
    attr_reader :transferInfo

    def initialize
      @value = '94070'
      @headers = {}
      @isActive = true
      @state = 'RINGING'
    end

    def answer; @state ='ANSWERED'; end
    def ask(text, options); AskResponse.new; end
    def callerID; '4155551212'; end
    def calledID; '4045551234'; end
    def callerName; 'Jason Goecke'; end
    def call(text, options); 'call response: ' + text.inspect; p options; end
    def conference(text); 'conference reponse: ' + text.inspect; end
    def getHeader(header); @headers[header]; end

    def hangup
      @isActive = false
      @state    = 'DISCONNECTED'
    end

    def sessionId; '1234'; end
    def log(text); text; end
    def meetme(text, *rest); "meetme: #{text.inspect}, #{rest.inspect}"; end
    def say(text, options); 'say response: text'; options; end
    def setHeader(header, value); @headers[header] = value; end
    def sipgetheader(calleridname); calleridname; end
    def startCallRecording(uri, options={}); nil ; end
    def stopCallRecording; nil; end
    def state; 'RINGING'; end
    def record(uri, options); true; end

    def transfer(destinations, options)
      @transferInfo = {:destinations => destinations, :options => options}
      return TropoResult.new
    end
  end

  class TropoResult
    attr_accessor :name

    def initialize
      @name = 'transfer'
    end
  end

  # Here for testing outside of Tropo, so we mock the $currentApp object
  class CurrentApp
    class GetApp
      def getApp
        'Application[http://hosting.tropo.com/49767/www/tropo-agi.rb:cus] ver(1.0.45500)'
      end
    end

    def initialize(app_id = 49767)
      @app_id = app_id
    end

    def self.app
      GetApp.new
    end

    def baseDir
      @cnt = 0 if @cnt.nil?

      # This is only for testing JT!!!
      case @app_id
      when 49767
        # Keep returning Windows format
        'c:\tropo_app_home\49767'
      when 49768
        # Then return the Linux format for the last test
        '/tropo_app_home/49768'
      else
        raise "Unknown application id!"
      end
    end
  end

  class IncomingCall
    def getHeaderMap
      map = java.util.HashMap.new
      map.put "kermit", "green"
      map.put "bigbird", "yellow"
      map
    end
  end
end
# Make these methods global.
Object.send(:include, Tropo)
@tropo_testing = true

# Ugly hack to make the tests run faster
class TropoAGItate
  class Commands
    def sleep(s)
      true
    end
  end
end


examples/adhearsion_dialplan.rb
# The main AGI entry point
tropo_agi {
  answer
  +hello_world
  hangup
}


# Invokes the native Tropo ask for Speech Recognition / ASR
# Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
ask_example {
  result = execute 'ask', { :prompt  => 'Please enter your zip code',
                            :choices => 'zipcode([5 DIGITS])' }.to_json
  # Get rid of the '200 result =' and then parse the JSON
  response = JSON.parse(result[11..-1])
  ahn_log.debug response
}

# We have lots of Asterisk sounds!
asterisk_sounds_example {
  play 'tt-monkeys'
}

# You may dial other SIP addresses, or telephone numbers as you like
# Formats may be:
#  SIP/
#  sip:
#  tel:+
# Related Tropo method: https://www.tropo.com/docs/scripting/transfer.htm
dial_example {
  dial 'sip:9991443146@sip.tropo.com'
}

hello_world {
  play 'tt-monkeys'
}

# Works as input should, only accepting DTMF
# Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
input_example {
  postal_code = input 5, :play         => 'Please enter your post code?',
                         :timeout      => 2

  ahn_log.postal_code.debug postal_code
  play "You entered"
  say_digits postal_code
}

# Find out if this is a Tropo session or an Asterisk one
is_tropo? {
  if type == 'TROPO'
    play "Yippeee! It is a Tropo call!"
  else
    play "No, this is a good old Asterisk call"
  end
}

# Related Tropo method: https://www.tropo.com/docs/scripting/say.htm
say_digits_example {
  say_digits '12345'
}

# Uses the native Tropo say method for Speech-Synthesis/TTS, will not play Asterisk sound files like play will
# Related Tropo method: https://www.tropo.com/docs/scripting/say.htm
say_example {
  execute 'say', { :prompt => 'I like to have weasels in my cloud.', :voice => 'simon' }.to_json
}

# If this is a Tropo call, then all of the SIP headers for the session are available
show_call_data {
  ahn_log.tropo_headers_str.debug tropo_headers
  tropo_headers = JSON.parse self.tropo_headers
  ahn_log.tropo_headers_hash.debug tropo_headers
  play "The content type is " + tropo_headers['Content-Type']
  +asterisk_sounds_example
}

# Monitor and Mixmonitor behave the same, may also be invoked as startCallRecording
# Related Tropo method: https://www.tropo.com/docs/scripting/startcallrecording.htm
monitor_example {
  play 'About to start call recording'
  execute 'monitor', { :uri                 => 'http://tropo-audiofiles-to-s3.heroku.com/post_audio_to_s3?filename=voicemail.mp3',
                       :format              => 'mp3',
                       :method              => 'POST',
                       :transcriptionOutURI => 'mailto:jsgoecke@voxeo.com' }.to_json
  play 'Call recording started!'
  play 'Thats it folks!'
  execute 'monitor_stop', ''
  play 'Recording stopped!'
}

# Related Tropo method: https://www.tropo.com/docs/scripting/record.htm
record_prompt_example {
  play 'Please record after the beep'
  record 'http://tropo-audiofiles-to-s3.heroku.com/post_audio_to_s3?filename=voicemail.mp3',
         :silence => 5,
         :maxduration => 120
}

# Allows you to set and retrieve variables on the session
# These also get passed as custom SIP headers, prepended with 'x-tropo' when you dial/transfer a call
variables_example {
  set_variable('foobar', 'green')
  ahn_log.debug get_variable('foobar')
}

# We have lots of Asterisk sounds!
asterisk_sounds_example {
  play 'tt-monkeys'
}

# Allows you to send standard DTMF digits
send_dtmf_example {
  dtmf '1234567890#*'
}

# Start menu example
# Related Tropo method: https://www.tropo.com/docs/scripting/ask.htm
menu_example {
  menu 'welcome', 'for spanish press 4',
       :timeout => 8.seconds, :tries => 3 do |link|
    link.shipment_status  1
    link.ordering         2
    link.representative   3
    link.spanish          4
    link.employee         500..599

    link.on_invalid { play 'invalid' }

    link.on_premature_timeout do |str|
      play 'sorry'
    end

    link.on_failure do
      play 'goodbye'
      hangup
    end
  end
}

shipment_status {
  play 'I surely do not know your shipment status.'
}

ordering {
  play 'Go somewhere else and order.'
}

representative {
  play 'No representatives here.'
}

spanish {
  play 'e 2 brutus?'
}

employee {
  play "The person at"
  say_digits extension
  play "went home for the day."
}
# End menu example
